tutorials/publish_subscribe/c/home_automation_publisher.c

```diff
@@ -0,0 +1,172 @@
+        if (retcode != DDS_RETCODE_OK) {
+            fprintf(stderr, "delete_participant error %d\n", retcode);
+            status = -1;
+        }
+    }
+
+    return status;
+}
+
+int publish_sensor(
+    char * sensor_name,
+    char * room_name)
+{
+
+    DDS_DomainParticipant *participant = NULL;
+    DDS_Publisher *publisher = NULL;
+    DDS_Topic *topic = NULL;
+    DDS_DataWriter *writer = NULL;
+    DeviceStatusDataWriter *DeviceStatus_writer = NULL;
+    const char *type_name = NULL;
+    DeviceStatus *device_status = NULL;
+    DDS_ReturnCode_t retcode;
+    struct DDS_Duration_t send_period = {2,0};
+    int i;
+
+    participant = DDS_DomainParticipantFactory_create_participant(
+            DDS_TheParticipantFactory,
+            0,
+            &DDS_PARTICIPANT_QOS_DEFAULT,
+            NULL /* listener */,
+            DDS_STATUS_MASK_NONE);
+
+    if (participant == NULL) {
+        fprintf(stderr, "create_participant error\n");
+        publisher_shutdown(participant);
+        return -1;
+    }
+
+    publisher = DDS_DomainParticipant_create_publisher(
+            participant,
+            &DDS_PUBLISHER_QOS_DEFAULT,
+            NULL /* listener */,
+            DDS_STATUS_MASK_NONE);
+
+    if (publisher == NULL) {
+        fprintf(stderr, "create_publisher error\n");
+        publisher_shutdown(participant);
+        return -1;
+    }
+
+    type_name = DeviceStatusTypeSupport_get_type_name();
+    retcode = DeviceStatusTypeSupport_register_type(
+            participant,
+            type_name);
+
+    if (retcode != DDS_RETCODE_OK) {
+        fprintf(stderr, "register_type error %d\n", retcode);
+        publisher_shutdown(participant);
+        return -1;
+    }
+
+    topic = DDS_DomainParticipant_create_topic(
+            participant,
+            "WindowStatus",
+            type_name,
+            &DDS_TOPIC_QOS_DEFAULT,
+            NULL /* listener */,
+            DDS_STATUS_MASK_NONE);
+
+    if (topic == NULL) {
+        fprintf(stderr, "create_topic error\n");
+        publisher_shutdown(participant);
+        return -1;
+    }
+
+    writer = DDS_Publisher_create_datawriter(
+            publisher,
+            topic,
+            &DDS_DATAWRITER_QOS_DEFAULT,
+            NULL /* listener */,
+            DDS_STATUS_MASK_NONE);
+    if (writer == NULL) {
+        fprintf(stderr, "create_datawriter error\n");
+        publisher_shutdown(participant);
+        return -1;
+    }
+
+    DeviceStatus_writer = DeviceStatusDataWriter_narrow(writer);
+    if (DeviceStatus_writer == NULL) {
+        fprintf(stderr, "DataWriter narrow error\n");
+        publisher_shutdown(participant);
+        return -1;
+    }
+
+    device_status = DeviceStatusTypeSupport_create_data_ex(DDS_BOOLEAN_TRUE);
+    if (device_status == NULL) {
+        fprintf(stderr, "DeviceStatusTypeSupport_create_data error\n");
+        publisher_shutdown(participant);
+        return -1;
+    }
+
+    device_status->sensor_name = sensor_name;
```

Comment:
this should copy the strings with `strncpy`; otherwise we're leaking the strings allocated by create_data above